File: algorithms/255.Verify_Preorder_Sequence_in_Binary_Search_Tree/Verify_Preorder_Sequence_in_Binary_Search_Tree.cc
// Verify Preorder Sequence in Binary Search Tree
class Solution {
public:
  bool verifyPreorder(vector<int>& preorder) {
    int p = INT_MIN;
    stack<int> s;
    for (auto x: preorder) {
      if (x < p)
        return false;
      while (! s.empty() && s.top() < x) {
        p = s.top();
        s.pop();
      }
      s.push(x);
    }
    return true;
  }
};
class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) {
        int low = INT_MIN;
        int i = -1;
        for(int j = 0; j < preorder.size(); j++) {
            if(preorder[j] < low) {
                return false;
            }

            while(i >= 0 && preorder[i] < preorder[j]) {
                low = preorder[i];
                i--;
            }

            i++;
            preorder[i] = preorder[j];
        }
        return true;
    }
};

File: algorithms/255.Verify_Preorder_Sequence_in_Binary_Search_Tree/Verify_Preorder_Sequence_in_Binary_Search_Tree.java
<<<<<<< HEAD
public class Solution {
    public boolean verifyPreorder(int[] preorder) {
        if (preorder == null || preorder.length < 3) {
            return true;
        }

        Stack<Integer> stack = new Stack<>();
        List<Integer> list = new ArrayList<>();

        for (int i : preorder) {
            if (!list.isEmpty() && i < list.get(list.size() - 1)) {
                return false;
            }

            while (!stack.isEmpty() && i > stack.peek()) {
                list.add(stack.pop());
            }
            stack.add(i);
        }

        return true;
    }
=======
public class Solution {
    public boolean verifyPreorder(int[] preorder) {
        if (preorder == null || preorder.length < 3) {
            return true;
        }

        Stack<Integer> stack = new Stack<>();
        List<Integer> list = new ArrayList<>();

        for (int i : preorder) {
            if (!list.isEmpty() && i < list.get(list.size() - 1)) {
                return false;
            }

            while (!stack.isEmpty() && i > stack.peek()) {
                list.add(stack.pop());
            }
            stack.add(i);
        }

        return true;
    }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
}